site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md

```diff
@@ -0,0 +1,548 @@
+
+As we saw above, changes enter the system by setting tracked properties. Recall that invoking `markAsUsed` bumps both the overall global clock value and the clock value for that property, and schedules a new render.[^coalescing] When the Glimmer VM re-renders, it can traverse the tree in a [depth-first search](https://medium.com/basecs/demystifying-depth-first-search-a7c14cccf056), comparing each frame’s current and cached clock values. If the clock value for a given frame hasn’t changed, nothing below it in the UI tree has changed, either—so we know we don’t need to re-render it. Checking whether that clock value has changed is literally just an integer equality check. At the nodes which *have* changed, the VM computes the new value and updates the DOM with the result.
+
+[^math-max]: There are some details about how it checks the tree and makes sure that it manages its internally state correctly, but it really is [using `Math.max`](https://github.com/glimmerjs/glimmer-vm/blob/e8e2fc6f39a60baac2b72c1a19aea9585b162c47/packages/%40glimmer/validator/lib/validators.ts#L130:L172).
+
+[^reactive-again]: or when using a “reactive function” via [the upcoming `invokeHelper` functionality][invoke-helper]
+
+[^coalescing]: The VM coalesces these bumps so if you set a bunch of values in response to user action or API responses or other inputs, it only triggers *one* re-render, not *many*.
+
+## Summary
+
+There are a handful of really delightful consequences of this system:
+
+- Re-renders are about as cheap as they possibly can be: all the state computations are simple integer math.
+- Intermediate, “derived” state gets computed on demand when the state it depends on changes—but with normal JavaScript semantics, without extra developer-facing boilerplate or end-user impact on performance.
+- It’s trivial to layer your own caching or memoization on top of these semantics if you need them, but you only pay for what you need.
+- All the “smarts” lives at the very edge of the system, in root state marked with `@tracked` and leaf values computed in reactive contexts like templates.
+
+Hopefully this has give you a good idea how autotracking works in general, and specifically how it simultaneously enables most of our code to be “just JavaScript” *and* gives us a very low-cost reactivity.
+
+:::callout
+
+If you’d like to see some of the details of how these pieces are implemented, check out [the video][walkthrough-video] of my conversation with Ember core team member and Glimmer VM contributor [Chris Garrett][cg] ([@pzuraq](https://github.com/pzuraq/)).
```

Comment:
Any chance you could link to my original autotracking blog post series, or talk from EmberConf? Not looking for exposure or anything, I just think those might be helpful as well 😛

Reply:
Yes! In fact, I meant to for the autotracking series but had forgotten about it for the EmberConf video. Great call.